### PR TITLE
skater

### PR DESCRIPTION
**Pull request by xf37**
add function skater and test_skater

### CODE REVIEW

File: spopt/region/skater.py
Comment (jGaboardi) [inline spopt/region/skater.py:285]: Move this to above the `if __name__ == "__main__":` line.

**Comment by xf37** [comment]
I got 2 failure checks about "cannot import name 'SKATER' from 'spopt'" when text_skater is running. Is it because the skater function has not been merged? Should I commit these two separately? Thanks in advance! @sjsrey

**Comment by jGaboardi** [comment]
add `sklearn` to [`requirements.txt`](https://github.com/pysal/spopt/blob/master/requirements.txt)

**Comment by jGaboardi** [comment]
These were the requirements for the original [`region`](https://github.com/pysal/region/blob/32861c36300e99e96ba15cfcc888d69a12046d2a/setup.py#L43) PySAL submodule. Also, it looks like `geopandas` is being used in the tests for the PR, but it is not in out current requirements. That should be added, as well.

File: requirements.txt
Comment (jGaboardi) [inline requirements.txt:7]: Ah! This should be `scikit-learn`! (See[ here](https://scikit-learn.org/stable/install.html#installation-instructions)). I am always confused about `scikit-learn`'s install name and import alias...

Comment (xf37) [reply requirements.txt]: Is it possible due to the [version of python](https://stackoverflow.com/questions/36404042/importerror-no-module-named-sklearn-python)?

Comment (jGaboardi) [reply requirements.txt]: It may be, but I am still not sure why the environment seems to not even attempt to download `scikit-learn`. See environment list [here](https://github.com/pysal/spopt/pull/65/checks?check_run_id=1059112512#step:3:832). I went through the `Run goanpeca/setup-miniconda@v1` process and did not even see any mention of `scikit-learn`. However, `geopandas` (which you just added) is being installed. This seems to be strange, but it will most likely be something small we overlooked... I have to work on some other stuff now, but maybe @knaaptime can give some advice?

Comment (jGaboardi) [reply requirements.txt]: [`conda-forge/scikit-learn`](https://github.com/conda-forge/scikit-learn-feedstock#current-build-status) is passing on all platforms, so it does not appear to be version/Python issue.

Comment (jGaboardi) [reply requirements.txt]: > This seems to be strange, but it will most likely be something small we overlooked... 

This ended up being true. It was an oversight of mine... see fe8dbcd

File: spopt/__init__.py
Comment (jGaboardi) [inline spopt/__init__.py:6]: Should we also expose `Spanning_Forest` here and in `region/__init__.py`?

File: spopt/tests/test_skater.py
Comment (jGaboardi) [inline spopt/tests/test_skater.py:6]: IMHO we should be using the `unittest`-style testing (as in [`esda`](https://github.com/pysal/esda/blob/e66098341ca8c8b31ceff2025edeefbf10df023f/esda/tests/test_moran.py#L12)), but I will defer to @sjsrey's judgment.

**Comment by jGaboardi** [comment]
There are also quite a few new lines of code in [`skater.py`](https://github.com/pysal/spopt/pull/65/checks?check_run_id=1061447495) that are not covered in [`test_skater.py`](https://github.com/pysal/spopt/pull/65/files#diff-a79f2c1c27338083576631e112a4d04a). We ideally should have more than one test, and have tests that are designed to fail in order to cover as many cases as possible (see example in [`spaghetti`](https://github.com/pysal/spaghetti/blob/be8fc7182da051ce20ea017c042cfcf134599bb6/spaghetti/tests/network_unittest_classes.py#L152)).

File: spopt/__init__.py
Comment (knaaptime) [reply spopt/__init__.py:6]: can we commit to pep8 from the get-go in this package? I'm having a tough time trying to convert some of the old packages (so `SpanningForest` if you want it) ?

Comment (knaaptime) [reply spopt/__init__.py:6]: also probably `Skater`...

Comment (jGaboardi) [reply spopt/__init__.py:6]: > can we commit to pep8 from the get-go in this package? I'm having a tough time trying to convert some of the old packages (so SpanningForest if you want it) ?

So do you mean:

```python
from region import MaxPHeuristic, RegionKMeansHeuristic, SKATER, w_to_g
```

instead of

```python
from .region import MaxPHeuristic, RegionKMeansHeuristic, SKATER, w_to_g
```
?

File: spopt/region/skater.py
Comment (knaaptime) [inline spopt/region/skater.py:14]: can we make this `SpanningForest` instead?

File: spopt/__init__.py
Comment (knaaptime) [reply spopt/__init__.py:6]: no i just mean all classes should be camel cased

Comment (jGaboardi) [reply spopt/__init__.py:6]: > no i just mean all classes should be camel cased

I really over-thought that one...

File: spopt/tests/test_skater.py
Comment (xf37) [reply spopt/tests/test_skater.py:7]: Thanks for pointing this out. I just followed the format of [test_maxp.py](https://github.com/pysal/spopt/blob/master/spopt/tests/test_maxp.py) and [test_region_k_means.py](https://github.com/pysal/spopt/blob/master/spopt/tests/test_region_k_means.py). I have not written this before. Do you know whether there are the unittest -style testing for these two methods? How could they pass the checks?  @jGaboardi

Comment (jGaboardi) [reply spopt/tests/test_skater.py:7]: @xf37 I don't mean to suggest that the way you wrote the test is wrong. It is a perfectly valid test. When I say "`unittest`-style", I simply mean with the `unittest` package in something like the format below. By creating a `TestSkater(unittest.TestCase)` instance and adding the `Skater` model as an attribute in `set_up()` we can reuse the model in more tests. In the current test we are only testing the values of `model.labels_`, but there are many of things we could (should?) be testing.

```python
import unittest

class TestSkater(unittest.TestCase):

    def set_up():
        self.pth = libpysal.examples.get_path("mexicojoin.shp")
        self.mexico = gpd.read_file(self.pth)
        self.mexico["count"] = 1
        self.attrs_name = [f"PCGDP{year}" for year in range(1950, 2010, 10)]
        self.w = libpysal.weights.Queen.from_dataframe(self.mexico)
        np.random.seed(123456)
        self.model = Skater(gdf=mexico, w=w, attrs_name=attrs_name)
        self.model.solve()

    def test_labels_(self):
        observed_labels = self.model.labels_
        known_labels = np.array(
            [
                0,
                0,
                1,
                2,
                2,
                1,
                1,
                1,
                1,
                1,
                3,
                2,
                1,
                1,
                1,
                1,
                4,
                1,
                1,
                1,
                1,
                1,
                0,
                0,
                0,
                1,
                1,
                1,
                1,
                0,
                1,
                1,
            ]
     )
     numpy.testing.assert_equal(observed_labels, known_labels)
```

File: spopt/region/skater.py
Comment (jGaboardi) [inline spopt/region/skater.py:340]: Let's go ahead and fill in the description here and the docstrings in `__init__`.

File: spopt/region/skater.py
Comment (jGaboardi) [inline spopt/region/skater.py:391]: Shouldn't users be able to pass in the keyword arguments to `SpanningForest.__init__()` here (e.g. `dissimilarity`)?

File: spopt/region/skater.py
Comment (jGaboardi) [inline spopt/region/skater.py:392]: Users should also be able to pass in keywords for `SpanningForest.fit()`  here.

File: spopt/tests/test_skater.py
Comment (xf37) [reply spopt/tests/test_skater.py:7]: Gotcha, thank you for explaining.

**Comment by xf37** [comment]
@jGaboardi Thank you for reviewing the PR! I have fixed all github CI test errors and would like to merge this PR. I may add more unit tests later.

File: spopt/region/skater.py
Comment (jGaboardi) [reply spopt/region/skater.py:415]: We should be passing in the keywords more programmatically. Something like the following, for example:

```python

class Skater(BaseSpOptHeuristicSolver):
    def __init__(..., ...., ...., spanning_forest_kwds=dict()):
        
        self.spanning_forest_kwds = spanning_forest_kwds
        
    
    def solve(self):
        """Solve the region k-means heuristic."""
        
        model = SpanningForest(**self.spanning_forest_kwds)
         

```

Then `Skater` can be instantiated like so:

```python
kwargs = {"center": numpy.median}
Skater(gdf=mexico, w=w, attrs_name=attrs_name, spanning_forest_kwds=kwargs)
```

File: spopt/region/skater.py
Comment (jGaboardi) [inline spopt/region/skater.py:419]: I am not sure there is a reason for having this last section since the tests are happening in `tests/test_skater.py`. Is there a reason for keeping it that I am not seeing?

**Review by jGaboardi — CHANGES_REQUESTED** [review]
I still have some concerns regarding the code before this gets merged (and we might want to double check if @sjsrey would like to give his review). I also still have concerns about the docstrings, but that can be beefed up and polished in a later PR.

File: spopt/region/skater.py
Comment (xf37) [reply spopt/region/skater.py]: Thanks, I just removed it.

**Comment by jGaboardi** [comment]
@xf37 Very nicely done! Let's wait until the end of the day if @sjsrey has time for a review then I'll merge this evening.

**Review by jGaboardi — APPROVED** [review]

**Pull request merged**
